### PR TITLE
Add DNS support for communication between VMs

### PR DESCRIPTION
**Pull request by kthomas**
This PR adds support for DNS name resolution between VMs running on the same nex host.

Nex node runs an embedded nameserver on an ephemeral port, which is provided to VMs by the agent via MMDS. The agent uses iptables to setup NAT OUTPUT chain rules for sending local DNS packets destined for port 53 (UDP or TCP) to the configured nameserver.

When a workload is deployed (or redeployed) to the Nex node, it's virtual interface is plugged in to the CNI bridge. If the workload is not configured to support DNS resolution, it still uses the CNI `ptp` plugin.

Workloads configured for name resolution and deployed in the `default` namespace are can be resolved by other VMs using `<workload name>`. Otherwise, workloads are resolved by `<namespace>.<workload name>`. It would be trivial to prevent name resolution of VMs deployed in one namespace by VMs deployed outside of that namespace, but that is not currently implemented.

DNS queries are handled by the nameserver if the hostname matches one of the configured workloads. Otherwise the nameserver acts as a recursive resolver, deferring to `127.0.0.53`.

This PR does not modify the local nameservers on the host, so name resolution only works within VMs at this time.

### CODE REVIEW

File: internal/node/config.go [hunk 1]
```diff
@@ -21,7 +21,7 @@ var (
 	defaultWorkloadTypes = []string{"elf", "v8", "wasm"}
 	defaultBinPath       = append([]string{"/usr/local/bin"}, filepath.SplitList(os.Getenv("PATH"))...)
 	// check the default cni bin path first, otherwise look in the rest of the PATH
-	defaultCNIBinPath = append([]string{"/opt/cni/bin"}, filepath.SplitList(os.Getenv("PATH"))...)
+	defaultCNIBinPath = []string{"/opt/cni/bin"}
```
[the remark this answers is not in the document]
Comment (kthomas) [reply internal/node/config.go:24]: We can workaround this issue by checking CNI plugin version or expected checksum.

**Comment by kthomas** [comment]
> * Adds a divergence between root file systems that support DNS and root file systems that don't, which increases our support boundaries

The divergence here is quite tiny, simply adding `iptables` and `sudo` packages to the root filesystem and the `nex` user as a `sudoer` on the VM.

**Comment by autodidaddict** [comment]
> > * Adds a dependency on the `bridge` CNI plugin that we don't currently use
> 
> The `bridge` plugin replaces the `ptp` plugin for workloads that are configured to support it.
> 
> > * Adds a divergence between root file systems that support DNS and root file systems that don't, which increases our support boundaries
> 
> The divergence here is quite tiny, simply adding `iptables` and `sudo` packages to the root filesystem and the `nex` user as a `sudoer` on the VM.

All workloads currently adhere to the same machine config template. So if one workload needs DNS support, then all of them have to have it enabled on the machine. While we have a feature scheduled to make templates that can be supplied during the deploy command, that's not there now.

The size of the divergence doesn't matter so much as the fact that now the `bridge` plugin needs to be installed on the node machine _just in case_ someone modifies the machine template to use bridge. and now precheck will have to be modified to vet the right bridge config.

**Comment by kthomas** [comment]
> > > * Adds a dependency on the `bridge` CNI plugin that we don't currently use
> > 
> > 
> > The `bridge` plugin replaces the `ptp` plugin for workloads that are configured to support it.
> > > * Adds a divergence between root file systems that support DNS and root file systems that don't, which increases our support boundaries
> > 
> > 
> > The divergence here is quite tiny, simply adding `iptables` and `sudo` packages to the root filesystem and the `nex` user as a `sudoer` on the VM.
> 
> All workloads currently adhere to the same machine config template. So if one workload needs DNS support, then all of them have to have it enabled on the machine. While we have a feature scheduled to make templates that can be supplied during the deploy command, that's not there now.
> 
> The size of the divergence doesn't matter so much as the fact that now the `bridge` plugin needs to be installed on the node machine _just in case_ someone modifies the machine template to use bridge. and now precheck will have to be modified to vet the right bridge config.
> 
> The impersonation attack vector is far higher up my list of concerns than the CNI plugins

Agreed, and we should probably support a much larger library of CNI plugins in the future, meaning many more should likely be present. This only matters for the host itself. The rootfs changes do not require awareness of any of these changes. The agent can be configured to conditionally enable the NAT OUTPUT chain rules.

**Comment by autodidaddict** [comment]
I think the issue is that right now we support all of the CNI plugins, by virtue of us not being opinionated there. Every time we enforce an opinion we limit the options and are on the hook to support it.

**Comment by kthomas** [comment]
> I think the issue is that right now we support all of the CNI plugins, by virtue of us not being opinionated there. Every time we enforce an opinion we limit the options and are on the hook to support it.

We actually only support the CNI plugins enabled in preflight-- this PR adds the `bridge` CNI plugin to that list. To support other CNI plugins we simply need to update the firecracker initialization code to read the CNI configuration provided for a workload and apply it instead of the default template. If a workload doesn't supply a CNI configuration, then still fall back to the default `ptp` template.

**Comment by autodidaddict** [comment]
Probably arguing semantics, but there's nothing stopping anyone from using a machine config that has plugins we don't verify via preflight. We only verify the _defaults_, not the entire world of supported scenarios.

**Comment by kthomas** [comment]
> Probably arguing semantics, but there's nothing stopping anyone from using a machine config that has plugins we don't verify via preflight. We only verify the _defaults_, not the entire world of supported scenarios.

We are probably both just being a bit pedantic at this point 😅

But currently we only pass the default template into the firecracker VM. Small amount of work to add CNI to the deploy request, read it and construct a dynamic `CNIConfiguration` -- or fallback to the default.

I am assuming we will want to keep the CNI plugins that can be applied at runtime to a workload in a whitelist of sorts though, to prevent nefarious things from happening to the host.

**Comment by autodidaddict** [comment]
Yeah, pedantic mode aside, some of this is discussed in #117

**Pull request closed without merging**